ember.md

```diff
@@ -0,0 +1,461 @@
+// good
+const { get, computed } = Ember;
+const {
+  get,
+  computed,
+} = Ember;
+
+export default Ember.Object.extend({
+  foo() {
+  },
+
+  actions: {
+    bar() {
+    },
+  },
+});
+```
+
+*Trailing commas are easier to maintain; one can add or remove properties without having to maintain the commas. Also produces cleaner git diffs.*
+
+### All functions must be named
+
+Any function declared with the `function` keyword must have a name. Functions declared with the es6 short "fat arrow" syntax do not need a name.
+
+This helps with debugging, as the name of the function shows up in the backtrace instead of `(anonymous function)`.
+
+When the function is used as part of a macro to define a property, we have a convention of using the property name prefixed with an underscore.
+
+```js
+// good
+
+things.map((thing) => get(thing, 'name'));
+
+fullName: computed('firstName', 'lastName', function _fullName() {
+  // Code
+}),
+
+// bad
+
+fullName: computed('firstName', 'lastName', function() {
+  // Code
+}),
+```
+
+## Organizing your modules
+
+Ordering a module's properties in a predictable manner will make it easier to
+scan.
+
+1. __Plain properties__
+
+   Start with properties that configure the module's behavior. Examples are
+   `tagName` and `classNames` on components and `queryParams` on controllers and
+   routes. Followed by any other simple properties, like default values for properties.
+
+2. __Single line computed property macros__
+
+   E.g. `alias`, `sort` and other macros. Start with service injections. If the
+   module is a model, then `attr` properties should be first, followed by
+   `belongsTo` and `hasMany`.
+
+3. __Multi line computed property functions__
+
+4. __Lifecycle hooks__
+
+   The hooks should be chronologically ordered by the order they are invoked in.
+
+5. __Functions__
+
+   Public functions first, internal functions after.
+
+6. __Actions__
+
+```js
+export default Component.extend({
+  // Plain properties
+  tagName: 'span',
+
+  // Single line CP
+  post: alias('myPost'),
+
+  // Multiline CP
+  authorName: computed('author.{firstName,lastName}', function _authorName() {
+    // code
+  }),
+
+  // Lifecycle hooks
+  didReceiveAttrs() {
+    this._super(...arguments);
+    // code
+  },
+
+  // Functions
+  someFunction() {
+    // code
+  },
+
+  actions: {
+    someAction() {
+      // Code
+    },
+  },
+});
+```
+
+### Override init
+
+Rather than using the object's `init` hook via `on`, override init and
+call `_super` with `...arguments`. This allows you to control execution
+order. [Don't Don't Override Init](https://dockyard.com/blog/2015/10/19/2015-dont-dont-override-init)
+
+## Models
+
+### Organization
+
+Models should be grouped as follows:
+
+* Attributes
+* Associations
+* Computed Properties
+
+Within each section, the attributes should be ordered alphabetically.
+
+```js
+// Good
+
+import Ember from 'ember';
+import DS from 'ember-data';
+
+const { computed } = Ember;
+
+const {
```

Comment:
Most of these examples have few properties, so they could be on a single line.

When we talked about this before, we discussed not making a hard and fast rule on this one. I feel like it is nice to use single lines when there are few properties to destructure, but it is also easy to add them when there are multiple lines (and easier to read when there are many). I also like splitting it up in to multiple lines when there are nested destructurings, such as:

```js
const {
  String: { w },
  RSVP: { resolve },
} = Ember;
```

I'll add a section for this.